wave_lang/kernel/wave/water_emitter.py

```
@@ -0,0 +1,364 @@
+    # Use textual parser to construct the Wave type.
+    type_text = _build_wave_type_text(wave_type_dict)
+    return ir.Type.parse(type_text, context=ctx)
+
+
+def create_operation_attributes(node_data):
+    """Create MLIR attributes from node data."""
+    attrs = {}
+
+    # Map common attributes
+    if "name" in node_data:
+        attrs["fx.name"] = ir.StringAttr.get(str(node_data["name"]))
+    if "op" in node_data:
+        attrs["fx.op"] = ir.StringAttr.get(str(node_data["op"]))
+    if "tkw_op_name" in node_data:
+        attrs["fx.tkw_op_name"] = ir.StringAttr.get(str(node_data["tkw_op_name"]))
+    if "subgraph" in node_data:
+        attrs["fx.subgraph"] = ir.StringAttr.get(str(node_data["subgraph"]))
+
+    # Map Python-specific attributes
+    for py_attr in (
+        "vector_shapes",
+        "reduction_dim",
+        "iter_idx",
+        "location",
+        "expanded_dims",
+        "scheduling_parameters",
+    ):
+        if py_attr in node_data:
+            attrs[f"py.{py_attr}"] = ir.StringAttr.get(str(node_data[py_attr]))
+
+    # Handle index attribute - store as string for now
+    if "index" in node_data:
+        index_str = node_data["index"]
+        attrs["py.index"] = ir.StringAttr.get(str(index_str))
+
+    return attrs
+
+
+def create_unregistered_operation(op_name, node_data, insertion_point):
+    """Fallback to create unregistered operation for unknown ops."""
+    attrs = create_operation_attributes(node_data)
+
+    return ir.Operation.create(
+        name=op_name,
+        attributes=attrs,
+        results=[],
+        operands=[],
+    )
+
+
+def main():
+    try:
+        data = json.load(sys.stdin)
+    except Exception as e:
+        print(f"FATAL: failed to read JSON: {e}", file=sys.stderr)
+        return 3
+
+    with ir.Context() as ctx:
+        # Only allow registered dialects/ops
+        try:
+            ctx.allow_unregistered_dialects = False
+        except Exception:
+            pass
```

Why is it fine to silence the exception here?